Source/Core/DolphinQt2/Debugger/MemoryViewWidget.cpp

```diff
@@ -0,0 +1,408 @@
+  verticalHeader()->hide();
+  verticalScrollBar()->setHidden(true);
+  setShowGrid(false);
+
+  setFont(Settings::Instance().GetDebugFont());
+
+  connect(&Settings::Instance(), &Settings::DebugFontChanged, this, &QWidget::setFont);
+  connect(&Settings::Instance(), &Settings::EmulationStateChanged, this, [this] { Update(); });
+  connect(this, &MemoryViewWidget::customContextMenuRequested, this,
+          &MemoryViewWidget::OnContextMenu);
+
+  setContextMenuPolicy(Qt::CustomContextMenu);
+
+  Update();
+}
+
+static int CalculateColumnCount(MemoryViewWidget::Type type)
+{
+  int column_count = 3;
+
+  switch (type)
+  {
+  case MemoryViewWidget::Type::ASCII:
+  case MemoryViewWidget::Type::U8:
+    column_count += 16;
+    break;
+  case MemoryViewWidget::Type::U16:
+    column_count += 8;
+    break;
+  case MemoryViewWidget::Type::U32:
+  case MemoryViewWidget::Type::Float32:
+    column_count += 4;
+    break;
+  }
+
+  return column_count;
+}
+
+void MemoryViewWidget::Update()
+{
+  clearSelection();
+
+  setColumnCount(CalculateColumnCount(m_type));
+
+  if (rowCount() == 0)
+    setRowCount(1);
+
+  setRowHeight(0, 24);
+
+  // Calculate (roughly) how many rows will fit in our table
+  int rows = std::round((height() / static_cast<float>(rowHeight(0))) - 0.25);
+
+  setRowCount(rows);
+
+  setColumnWidth(0, 24);
+
+  for (int i = 0; i < rows; i++)
+  {
+    setRowHeight(i, 24);
+
+    u32 addr = m_address - ((rowCount() / 2) * 16) + i * 16;
+
+    auto* bp_item = new QTableWidgetItem;
+    bp_item->setFlags(Qt::ItemIsEnabled);
+    bp_item->setData(Qt::UserRole, addr);
+
+    if (PowerPC::memchecks.OverlapsMemcheck(addr, 16))
+      bp_item->setBackground(Qt::red);
+
+    auto* addr_item = new QTableWidgetItem(QStringLiteral("%1").arg(addr, 8, 16, QLatin1Char('0')));
+
+    addr_item->setData(Qt::UserRole, addr);
+    addr_item->setFlags(Qt::ItemIsEnabled | Qt::ItemIsSelectable);
+
+    setItem(i, 0, bp_item);
+    setItem(i, 1, addr_item);
+
+    if (addr == m_address)
+      addr_item->setSelected(true);
+
+    if (Core::GetState() != Core::State::Paused || !PowerPC::HostIsRAMAddress(addr))
+    {
+      for (int c = 2; c < columnCount(); c++)
+      {
+        auto* item = new QTableWidgetItem(QStringLiteral("-"));
+        item->setFlags(Qt::ItemIsEnabled);
+        item->setData(Qt::UserRole, addr);
+
+        setItem(i, c, item);
+      }
+
+      continue;
+    }
+
+    auto* description_item =
+        new QTableWidgetItem(QString::fromStdString(PowerPC::debug_interface.GetDescription(addr)));
+
+    description_item->setForeground(Qt::blue);
+    description_item->setFlags(Qt::ItemIsEnabled);
+
+    setItem(i, columnCount() - 1, description_item);
+
+    switch (m_type)
+    {
+    case Type::U8:
+      for (int c = 0; c < 16; c++)
+      {
+        auto* hex_item = new QTableWidgetItem;
+
+        if (PowerPC::HostIsRAMAddress(addr + c * sizeof(u8)))
+        {
+          u8 value = PowerPC::HostRead_U8(addr + c * sizeof(u8));
```

Comment:
You just use it twice in this case, but i think introducing a variable for `addr + c * sizeof(u8)` (and also the others in the following cases) would help readability. Might just be me though.

Reply:
If it'd be more than two uses I'd agree. But in this case it doesn't really matter that much IMHO.